doc/bsetroot.1
.\"
.\" Man page for Bsetroot
.\"
.\" Copyright (c) 2000 by Wilbert Berendsen <wbsoft@xs4all.nl>
.\"
.TH bsetroot 1 "June 16th, 2000" "0.60.3"
.SH NAME
bsetroot \- Openbox utility to change root window appearance
.SH SYNOPSIS
.BR bsetroot " \-help"
.br
.B bsetroot
.RI "[ \-display " display " ] \-solid " color
.br
.B bsetroot
.RI "[ \-display " display " ] \-mod " "x y" " \-fg " color " \-bg " color
.br
.B bsetroot
.RI "[ \-display " display " ] \-gradient " texture " \-from " color " \-to " color
.SH DESCRIPTION
Bsetroot is a utility that can control the appearance of the root window in
three ways: Either give it a solid color, or write a two color modula pattern
to it, or render a gradient texture, based on two different colors.
.PP
Bsetroot resembles
.IR xsetroot (1)
in this functionality but it supports multiple screen displays, and gradient
textures the same way as Openbox does.
It doesn't handle cursors etc.
Bsetroot is part of the Openbox package.
.SH OPTIONS
Bsetroot operates in three ways, you must choose one of the first 3 options:
.TP
.BI \-solid " color"
Sets the root window to specified color.
.TP
.BI \-mod " x y"
Creates a modula pattern. You must specify
.BR \-bg " and " \-fg
colors.
.TP
.BI \-gradient " texturestring"
Renders the specified texture string to the root window.
For possible texture strings, please refer to
.IR openbox (1).
You must also specify both a
.BR \-from " and a " \-to
color.
.TP
.BI \-display " display"
Tells Bsetroot to connect to the specified display.
.TP
.BI "\-bg, \-background " color
Background color.
Needed for
.B \-mod
patterns.
.TP
.BI "\-fg, \-foreground " color
Foreground color.
Needed for
.B \-mod
patterns.
.TP
.BI \-from " color"
Start color for rendering textures.
Needed for
.B \-gradient
operation mode.
.TP
.BI \-to " color"
Ending color for rendering textures.
Needed for
.B \-gradient
operation mode.
.TP
.B \-help
Prints version info and short help text.
.SH AUTHOR
Bsetroot was written and maintained by Brad Hughes
.nh \" hyphenation off
(blackbox@alug.org)
.hy \" on again
and Jeff Raven
.nh
(jraven@psu.edu).
.hy
.SH SEE ALSO
.IR openbox (1)


CHANGELOG
3.7:
  * Update translations.
  * Fix ShapeInput shape mask propagation (clickthrough windows).
  * Add fullscreen, focusable conditions to If and ForEach actions.
  * Add <none> to ForEach, ran if there were no matches.
    (The actions in the <else> tag are run for every checked client.)
  * Allow ignoring repeated keyboard events with XKB, using the repeat="no"
    attribute. Useful to not accidentally close every window if your F4 key
    gets stuck.
  * Fix a crash that was exposed by glib removing their arena allocator.
  * Add the FocusFallback action which does the same as the Unfocus action
    but with a more logical name. The Unfocus action may change to unfocus the
    window in a future release (and not focus any other window in its place).
  * Various small bugfixes.
  * Added focus.cycle.bg.color and focus.cycle.border.color theme options.

3.6.1:
  * Bump ABI version of libobrender that should have been done in 3.6

3.6:
  * Added various new abilities to If action, including the <query> tag,
    matching class, name, role and window type.
  * Added ForEach action.
  * Restore the <center> option for window placement.
  * New GrowToFill action.
  * Updated Portuguese and Polish translations.
  * New Galician translation.
  * Added strict option to ToggleShowDesktop action.
  * Speed optimization in rendering code.
  * New <position> tag for ShowMenu action allows placing menus other
    than at the mouse cursor.
  * Fix some corner cases of relative position to work correctly.
  * Allow specifying multiple mousebinds in one tag as we already do
    for keybinds (separated by spaces, as "W-Return M-S-F3" for example).
  * Added option rebindOnMappingNotify option, on by default, which
    controls if we update binds when the keyboard layout changes at runtime.
  * The resize action now only considers the part of the client that is onscreen
    for purposes of picking an action to perform.
  * Make desktop actions non-interactive (means you can have more actions bound
    to a key after them, and the dialog disappears on a timer instead of when
    releasing the modifier keys).
  * obxprop is now not pathologically slow when printing very long arrays.
  * Fix menu accelerator underlines sometimes not showing up.
  * Fix so you can click menu entries before the menuHideDelay expires if it's
    a new click.
  * Don't readd the X event fd to the poll set every time we reconfigure.
  * Fix some specialized button bitmaps not loading from the correct files.
    (toggled + pressed or hover would just be the base toggled image).
  * Various small bugfixes.

3.5.2:
  * Fix crash in theme rendering code, and in theme parsing code.
  * Maximize windows without borders to fill the whole screen, don't leave
    a one-pixel empty space. (Bug #5996)

3.5.1:
  * New translations: Afrikaans, Belarusian, Interlingua, Hebrew, Romanian,
    Greek.
  * Updated translations: Italian, Serbian, Spanish, Czech, Hungarian, Turkish,
    German, Arabic, Polish, Dutch, Lithuanian, Portuguese, Estonian.
  * SVG icon support. This is optional, depending on librsvg being installed.
  * Allow application rules to control window size with a new <size> tag.
  * Allow application rules to pick a monitor for new windows without forcing
    a position.
  * Allow non-interactive focus cycling with a new <interactive> tag.
  * New LeastOverlap window placement policy replaces the old default
    behaviour. It finds a place on a given monitor that overlaps as few
    windows as possible. Contributed by Ian Zimmerman.
  * Improved Xinerama behaviour.
  * Correctly interface with latest gnome-session for Gnome/Openbox X sessions.
  * Allow third-party control of window opacity in compositing managers.
  * Improved themeing options. Contributed by Dave Foster.
  * Add <monitor>, <title type="regex">, <title type="exact"> and
    <activedesktop> options to If action.
  * Addresses bugs #4661, #5506, #5186, #5758, #5410, #5228, #5277, #5731,
    #5746, #5737, #5419, #5721, #5711, #5385, #5500, #4992, #5443, #5518,
    #5444, #4782, #5237, #5228, #5173, #5203, #5246, #5180, #5179, #5150,
    #5132, #4937, #4889, #5253, #3769, #5819, #5811, #5081, #5426 among others.

3.5.0:
  * New alt-tab dialog shows windows in a vertical list.
  * Improved Xinerama support.
  * Allow icons in menus.
  * Theme options for prompt dialogs (osd.button.unpressed.*,
    osd.button.pressed.*, osd.button.focused.*)
  * Addresses bug #4877, #4596, #4617, #4752, #4663, #4662, #4586, #2319,
    #4341, #4519, #4543, #4503, #4355, #4072, #3702, #4284
  * Lots of additional bug fixes and performance improvements.

3.4.11.2:
  * Updated Estonian and Portuguese translations.
  * Fix a rare crash involving moving fullscreen windows to different monitors
  * Fix a more common crash involving pressing right in a menu

3.4.11.1:
  * Updated Polish translation.
  * Fixed bug #4519 (Incorrect focus in reused windows).
  * Lower the default submenu open/hide delay from 200ms to 100ms.
  * Fix some more problems with gnome integration.
  * Update Clearlooks theme.
  * Some other small fixes.

3.4.11:
  * Update Hungarian, Japanese, and Latvian translations.
  * Make xdg-autostart use the OPENBOX environment by default, so you can use
    OnlyShowIn=OPENBOX in an autostart .desktop and it will work as expected.
  * Don't close the menu when you hold control and execute something.
  * Fix bug #4503 (Adjust who shows up in the Alt-Tab list using SKIP_TASKBAR).
  * Fix flickering window when moving maximized window between monitors of
    different sizes.
  * Fix bug #4355 (Allow multiple escaped _'s in a menu label and allow
    a menu shortcut to come after an escaped _).
  * Remember the maximized state of a window when it goes fullscreen, and
    restore it when leaving fullscreen state.
  * Fix bug #4072 (Openbox is stopped by terminal applications writing to
    stdout).
  * Fix bug #4492 (Mistake in openbox-gnome-session check while setting up).

util/bsetroot.cc
#ifdef    HAVE_CONFIG_H
#  include "../config.h"
#endif // HAVE_CONFIG_H

#ifdef    HAVE_STDLIB_H
#  include <stdlib.h>
#endif // HAVE_STDLIB_H

#ifdef    HAVE_STRING_H
#  include <string.h>
#endif // HAVE_STRING_H

#ifdef    HAVE_STDIO_H
#  include <stdio.h>
#endif // HAVE_STDIO_H

#include "../src/i18n.h"
#include "../src/BaseDisplay.h"
#include "../src/Image.h"
#include "bsetroot.h"


bsetroot::bsetroot(int argc, char **argv, char *dpy_name)
  : BaseDisplay(argv[0], dpy_name)
{
  grad = fore = back = (char *) 0;

  Bool mod = False, sol = False, grd = False;
  int mod_x = 0, mod_y = 0;

  for (int i = 1; i < argc; i++) {
    if (! strcmp("-help", argv[i])) {
      usage();
    } else if ((! strcmp("-fg", argv[i])) ||
               (! strcmp("-foreground", argv[i])) ||
               (! strcmp("-from", argv[i]))) {
      if ((++i) >= argc) usage(1);

      fore = argv[i];
    } else if ((! strcmp("-bg", argv[i])) ||
               (! strcmp("-background", argv[i])) ||
               (! strcmp("-to", argv[i]))) {
      if ((++i) >= argc) usage(1);

      back = argv[i];
    } else if (! strcmp("-solid", argv[i])) {
      if ((++i) >= argc) usage(1);

      fore = argv[i];
      sol = True;
    } else if (! strcmp("-mod", argv[i])) {
      if ((++i) >= argc) usage();

      mod_x = atoi(argv[i]);

      if ((++i) >= argc) usage();

      mod_y = atoi(argv[i]);

      if (mod_x < 1) mod_x = 1;
      if (mod_y < 1) mod_y = 1;

      mod = True;
    } else if (! strcmp("-gradient", argv[i])) {
      if ((++i) >= argc) usage();

      grad = argv[i];
      grd = True;
    } else if (! strcmp("-display", argv[i])) {
      // -display passed through tests ealier... we just skip it now
      i++;
    } else
      usage();
  }

  if ((mod + sol + grd) != True) {
    fprintf(stderr,
	    i18n->getMessage(bsetrootSet, bsetrootMustSpecify,
		  "%s: error: must specify one of: -solid, -mod, -gradient\n"),
	    getApplicationName());
    
    usage(2);
  }
  
  img_ctrl = new BImageControl*[getNumberOfScreens()];
  for (int i = 0; i < getNumberOfScreens(); i++)
    img_ctrl[i] = new BImageControl(*this, *getScreenInfo(i), True);

  if (sol && fore) solid();
  else if (mod && mod_x && mod_y && fore && back) modula(mod_x, mod_y);
  else if (grd && grad && fore && back) gradient();
  else usage();
}


bsetroot::~bsetroot(void) {
  XSetCloseDownMode(getXDisplay(), RetainPermanent);

  XKillClient(getXDisplay(), AllTemporary);

  for (int i = 0; i < getNumberOfScreens(); i++)
    delete img_ctrl[i];

  delete [] img_ctrl;
}

// adapted from wmsetbg
void bsetroot::setPixmapProperty(int screen, Pixmap pixmap) {
  static Atom rootpmap_id = None;
  Atom type;
  int format;
  unsigned long length, after;
  unsigned char *data;
  int mode;
  
  if (rootpmap_id == None)
    rootpmap_id = XInternAtom(getXDisplay(), "_XROOTPMAP_ID", False);

  XGrabServer(getXDisplay());
  
  /* Clear out the old pixmap */
  XGetWindowProperty(getXDisplay(), getScreenInfo(screen)->getRootWindow(),
		     rootpmap_id, 0L, 1L, False, AnyPropertyType,
		     &type, &format, &length, &after, &data);
  
  if ((type == XA_PIXMAP) && (format == 32) && (length == 1)) {
    XKillClient(getXDisplay(), *((Pixmap *)data));
    XSync(getXDisplay(), False);
    mode = PropModeReplace;
  } else {
    mode = PropModeAppend;
  }
  if (pixmap) {
    XChangeProperty(getXDisplay(), getScreenInfo(screen)->getRootWindow(),
		    rootpmap_id, XA_PIXMAP, 32, mode,
		    (unsigned char *) &pixmap, 1);
  } else {
    XDeleteProperty(getXDisplay(), getScreenInfo(screen)->getRootWindow(),
		    rootpmap_id);
  }
  
  XUngrabServer(getXDisplay());
  XFlush(getXDisplay());
}


// adapted from wmsetbg
Pixmap bsetroot::duplicatePixmap(int screen, Pixmap pixmap,
				 int width, int height) {
  XSync(getXDisplay(), False);

  Pixmap copyP = XCreatePixmap(getXDisplay(),
			       getScreenInfo(screen)->getRootWindow(),
			       width, height,
			       DefaultDepth(getXDisplay(), screen));
  XCopyArea(getXDisplay(), pixmap, copyP, DefaultGC(getXDisplay(), screen),
	    0, 0, width, height, 0, 0);
  XSync(getXDisplay(), False);

  return copyP;
}


void bsetroot::solid(void) {
  for (int screen = 0; screen < getNumberOfScreens(); screen++) {
    BColor c;

    img_ctrl[screen]->parseColor(&c, fore);
    if (! c.isAllocated()) c.setPixel(BlackPixel(getXDisplay(), screen));

    XSetWindowBackground(getXDisplay(), getScreenInfo(screen)->getRootWindow(),
                         c.getPixel());
    XClearWindow(getXDisplay(), getScreenInfo(screen)->getRootWindow());

    Pixmap pixmap = XCreatePixmap(getXDisplay(),
				  getScreenInfo(screen)->getRootWindow(),
				  8, 8, DefaultDepth(getXDisplay(), screen));
    XSetForeground(getXDisplay(), DefaultGC(getXDisplay(), screen),
		   c.getPixel());
    XFillRectangle(getXDisplay(), pixmap, DefaultGC(getXDisplay(), screen),
		   0, 0, 8, 8);

    setPixmapProperty(screen, duplicatePixmap(screen, pixmap, 8, 8));

    XFreePixmap(getXDisplay(), pixmap);
  }
}


void bsetroot::modula(int x, int y) {
  char data[32];
  long pattern;

  int screen, i;

  for (pattern = 0, screen = 0; screen < getNumberOfScreens(); screen++) {
    for (i = 0; i < 16; i++) {
      pattern <<= 1;
      if ((i % x) == 0)
        pattern |= 0x0001;
    }

    for (i = 0; i < 16; i++) {
      if ((i %  y) == 0) {
        data[(i * 2)] = (char) 0xff;
        data[(i * 2) + 1] = (char) 0xff;
      } else {
        data[(i * 2)] = pattern & 0xff;
        data[(i * 2) + 1] = (pattern >> 8) & 0xff;
      }
    }

    BColor f, b;
    GC gc;
    Pixmap bitmap;
    XGCValues gcv;

    bitmap =
      XCreateBitmapFromData(getXDisplay(),
                            getScreenInfo(screen)->getRootWindow(), data,
                            16, 16);

    img_ctrl[screen]->parseColor(&f, fore);
    img_ctrl[screen]->parseColor(&b, back);

    if (! f.isAllocated()) f.setPixel(WhitePixel(getXDisplay(), screen));
    if (! b.isAllocated()) b.setPixel(BlackPixel(getXDisplay(), screen));

    gcv.foreground = f.getPixel();
    gcv.background = b.getPixel();

    gc = XCreateGC(getXDisplay(), getScreenInfo(screen)->getRootWindow(),
                   GCForeground | GCBackground, &gcv);

    Pixmap pixmap = XCreatePixmap(getXDisplay(),
				  getScreenInfo(screen)->getRootWindow(),
				  16, 16, getScreenInfo(screen)->getDepth());

    XCopyPlane(getXDisplay(), bitmap, pixmap, gc,
               0, 0, 16, 16, 0, 0, 1l);
    XSetWindowBackgroundPixmap(getXDisplay(),
                               getScreenInfo(screen)->getRootWindow(),
                               pixmap);
    XClearWindow(getXDisplay(), getScreenInfo(screen)->getRootWindow());

    setPixmapProperty(screen,
		      duplicatePixmap(screen, pixmap, 16, 16));

    XFreeGC(getXDisplay(), gc);
    XFreePixmap(getXDisplay(), bitmap);

    if (! (getScreenInfo(screen)->getVisual()->c_class & 1))
      XFreePixmap(getXDisplay(), pixmap);
  }
}


void bsetroot::gradient(void) {
  for (int screen = 0; screen < getNumberOfScreens(); screen++) {
    BTexture texture;
    img_ctrl[screen]->parseTexture(&texture, grad);
    img_ctrl[screen]->parseColor(texture.getColor(), fore);
    img_ctrl[screen]->parseColor(texture.getColorTo(), back);

    if (! texture.getColor()->isAllocated())
      texture.getColor()->setPixel(WhitePixel(getXDisplay(), screen));
    if (! texture.getColorTo()->isAllocated())
      texture.getColorTo()->setPixel(BlackPixel(getXDisplay(), screen));

    Pixmap pixmap =
      img_ctrl[screen]->renderImage(getScreenInfo(screen)->size().w(),
                                    getScreenInfo(screen)->size().h(),
                                    &texture);

    XSetWindowBackgroundPixmap(getXDisplay(),
                               getScreenInfo(screen)->getRootWindow(),
                               pixmap);
    XClearWindow(getXDisplay(), getScreenInfo(screen)->getRootWindow());

    setPixmapProperty(screen,
		      duplicatePixmap(screen, pixmap,
				      getScreenInfo(screen)->size().w(), 
				      getScreenInfo(screen)->size().h()));
      
    if (! (getScreenInfo(screen)->getVisual()->c_class & 1)) {
      img_ctrl[screen]->removeImage(pixmap);
    }
  }
}


void bsetroot::usage(int exit_code) {
  fprintf(stderr,
          i18n->
	  getMessage(bsetrootSet, bsetrootUsage,
	             "%s 2.0 : (c) 1997-1999 Brad Hughes\n\n"
		     "  -display <string>        display connection\n"
		     "  -mod <x> <y>             modula pattern\n"
		     "  -foreground, -fg <color> modula foreground color\n"
		     "  -background, -bg <color> modula background color\n\n"
		     "  -gradient <texture>      gradient texture\n"
		     "  -from <color>            gradient start color\n"
		     "  -to <color>              gradient end color\n\n"
		     "  -solid <color>           solid color\n\n"
		     "  -help                    print this help text and exit\n"),
	  getApplicationName());
  
  exit(exit_code);
}


int main(int argc, char **argv) {
  char *display_name = (char *) 0;
  
  NLSInit("openbox.cat");
  
  for (int i = 1; i < argc; i++) {
    if (! strcmp(argv[i], "-display")) {
      // check for -display option
      
      if ((++i) >= argc) {
        fprintf(stderr,
		i18n->getMessage(mainSet, mainDISPLAYRequiresArg,
				 "error: '-display' requires an argument\n"));
	
        ::exit(1);
      }
      
      display_name = argv[i];
    }
  }
  
  bsetroot app(argc, argv, display_name);
  
  return 0;
}
